File: src/arithmetic_mixed_numbers_medium.py
"""
arithmetic - mixed_numbers (medium)
Generated: 2026-02-22T03:41:41.674311
"""

from problem_utils import *

def generate():
    # Choose operation type
    op_type = choice(['add', 'subtract', 'multiply', 'divide'])
    
    # Generate random mixed numbers - construct from clean parts
    # First mixed number
    whole1 = randint(1, 5)
    numer1 = randint(1, 7)
    denom1 = randint(numer1 + 1, 10)  # Ensure proper fraction
    
    # Second mixed number
    whole2 = randint(1, 5)
    numer2 = randint(1, 7)
    denom2 = randint(numer2 + 1, 10)  # Ensure proper fraction
    
    # Convert to improper fractions for calculation
    improper1 = Rational(whole1 * denom1 + numer1, denom1)
    improper2 = Rational(whole2 * denom2 + numer2, denom2)
    
    # Perform operation
    if op_type == 'add':
        result = improper1 + improper2
        op_symbol = '+'
        op_word = 'Add'
    elif op_type == 'subtract':
        # Ensure positive result
        if improper1 < improper2:
            improper1, improper2 = improper2, improper1
            whole1, whole2 = whole2, whole1
            numer1, numer2 = numer2, numer1
            denom1, denom2 = denom2, denom1
        result = improper1 - improper2
        op_symbol = '-'
        op_word = 'Subtract'
    elif op_type == 'multiply':
        result = improper1 * improper2
        op_symbol = '\\times'
        op_word = 'Multiply'
    else:  # divide
        result = improper1 / improper2
        op_symbol = '\\div'
        op_word = 'Divide'
    
    # Format mixed numbers for display
    mixed1_latex = f"{whole1}\\frac{{{numer1}}}{{{denom1}}}"
    mixed2_latex = f"{whole2}\\frac{{{numer2}}}{{{denom2}}}"
    
    # Convert result to mixed number if improper
    result_whole = int(result)
    result_remainder = result - result_whole
    
    if result_remainder == 0:
        answer_latex = str(result_whole)
        answer_value = result_whole
    elif abs(result) < 1:
        # Proper fraction
        answer_latex = f"\\frac{{{result.p}}}{{{result.q}}}"
        answer_value = result
    else:
        # Mixed number
        if result < 0:
            result_whole = int(result) - 1
            result_remainder = result - result_whole
        answer_whole = abs(result_whole)
        frac_part = abs(result_remainder)
        if result < 0:
            answer_latex = f"-{answer_whole}\\frac{{{frac_part.p}}}{{{frac_part.q}}}"
        else:
            answer_latex = f"{answer_whole}\\frac{{{frac_part.p}}}{{{frac_part.q}}}"
        answer_value = result
    
    # Build solution steps
    improper1_latex = f"\\frac{{{improper1.p}}}{{{improper1.q}}}"
    improper2_latex = f"\\frac{{{improper2.p}}}{{{improper2.q}}}"
    
    solution_steps = [
        f"Convert mixed numbers to improper fractions:",
        f"${mixed1_latex} = {improper1_latex}$ and ${mixed2_latex} = {improper2_latex}$"
    ]
    
    if op_type in ['add', 'subtract']:
        # Find LCD
        from math import gcd
        lcm = (denom1 * denom2) // gcd(denom1, denom2)
        mult1 = lcm // denom1
        mult2 = lcm // denom2
        new_numer1 = improper1.p * mult1
        new_numer2 = improper2.p * mult2
        
        solution_steps.append(f"Find common denominator (LCD = {lcm}):")
        solution_steps.append(f"${improper1_latex} = \\frac{{{new_numer1}}}{{{lcm}}}$ and ${improper2_latex} = \\frac{{{new_numer2}}}{{{lcm}}}$")
        
        if op_type == 'add':
            solution_steps.append(f"Add the numerators: $\\frac{{{new_numer1} + {new_numer2}}}{{{lcm}}} = \\frac{{{result.p}}}{{{result.q}}}$")
        else:
            solution_steps.append(f"Subtract the numerators: $\\frac{{{new_numer1} - {new_numer2}}}{{{lcm}}} = \\frac{{{result.p}}}{{{result.q}}}$")
    elif op_type == 'multiply':
        solution_steps.append(f"Multiply numerators and denominators:")
        solution_steps.append(f"$\\frac{{{improper1.p} \\times {improper2.p}}}{{{improper1.q} \\times {improper2.q}}} = \\frac{{{result.p}}}{{{result.q}}}$")
    else:  # divide
        solution_steps.append(f"Multiply by the reciprocal:")
        solution_steps.append(f"$\\frac{{{improper1.p}}}{{{improper1.q}}} \\times \\frac{{{improper2.q}}}{{{improper2.p}}} = \\frac{{{result.p}}}{{{result.q}}}$")
    
    solution_steps.append(f"Convert back to mixed number: ${answer_latex}$")
    
    # Set difficulty based on operation
    if op_type in ['add', 'subtract']:
        diff = (1300, 1500)
    else:
        diff = (1400, 1600)
    
    return problem(
        question=f"${mixed1_latex} {op_symbol} {mixed2_latex}$",
        answer=answer_value,
        difficulty=diff,
        topic="arithmetic/mixed_numbers",
        solution=steps(*solution_steps),
        answer_type="numeric"
    )

emit(generate())

File: src/arithmetic_mixed_numbers_medium.jl
# arithmetic - mixed_numbers (medium)
# Generated: 2026-03-08T19:49:00.110713

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("arithmetic/mixed_numbers")
    
    prob_type = rand(1:6)
    
    if prob_type == 1
        # Addition of mixed numbers requiring regrouping
        whole1 = randint(2, 8)
        whole2 = randint(2, 8)
        num1 = randint(1, 7)
        num2 = randint(1, 7)
        denom = choice([3, 4, 5, 6, 8, 10, 12])
        
        # Ensure addition requires regrouping
        num1 = randint(denom ÷ 2, denom - 1)
        num2 = randint(denom ÷ 2, denom - 1)
        
        sum_num = num1 + num2
        extra_wholes = sum_num ÷ denom
        final_num = sum_num % denom
        final_whole = whole1 + whole2 + extra_wholes
        
        # Simplify if needed
        g = gcd(final_num, denom)
        final_num_simp = final_num ÷ g
        final_denom = denom ÷ g
        
        answer = final_num_simp == 0 ? "$final_whole" : "$(final_whole) $(final_num_simp)/$(final_denom)"
        
        problem(
            question="Add and simplify: \$$(whole1)\\frac{$(num1)}{$(denom)} + $(whole2)\\frac{$(num2)}{$(denom)}\$",
            answer=answer,
            difficulty=(1200, 1400),
            solution=steps(
                "Add whole numbers: \$$(whole1) + $(whole2) = $(whole1 + whole2)\$",
                "Add fractions: \$\\frac{$(num1)}{$(denom)} + \\frac{$(num2)}{$(denom)} = \\frac{$(sum_num)}{$(denom)}\$",
                "Convert improper fraction: \$\\frac{$(sum_num)}{$(denom)} = $(extra_wholes)\\frac{$(final_num)}{$(denom)}\$",
                "Combine: \$$(whole1 + whole2) + $(extra_wholes)\\frac{$(final_num)}{$(denom)} = $(final_whole)\\frac{$(final_num)}{$(denom)}\$",
                g > 1 ? "Simplify: \$$(answer)\$" : "Answer: \$$(answer)\$"
            ),
            time=120
        )
        
    elseif prob_type == 2
        # Subtraction of mixed numbers requiring borrowing
        whole1 = randint(5, 12)
        whole2 = randint(2, whole1 - 2)
        denom = choice([4, 5, 6, 8, 10, 12])
        num1 = randint(1, denom - 1)
        num2 = randint(num1 + 1, denom - 1)  # Force borrowing
        
        # Convert to improper fractions
        imp1 = whole1 * denom + num1
        imp2 = whole2 * denom + num2
        result = imp1 - imp2
        
        final_whole = result ÷ denom
        final_num = result % denom
        
        g = gcd(final_num, denom)
        final_num_simp = final_num ÷ g
        final_denom = denom ÷ g
        
        answer = final_num_simp == 0 ? "$final_whole" : "$(final_whole) $(final_num_simp)/$(final_denom)"
        
        problem(
            question="Subtract and simplify: \$$(whole1)\\frac{$(num1)}{$(denom)} - $(whole2)\\frac{$(num2)}{$(denom)}\$",
            answer=answer,
            difficulty=(1300, 1500),
            solution=steps(
                "Convert to improper fractions: \$\\frac{$(imp1)}{$(denom)} - \\frac{$(imp2)}{$(denom)}\$",
                "Subtract: \$\\frac{$(result)}{$(denom)}\$",
                "Convert to mixed number: \$$(final_whole)\\frac{$(final_num)}{$(denom)}\$",
                g > 1 ? "Simplify: \$$(answer)\$" : "Answer: \$$(answer)\$"
            ),
            time=150
        )
        
    elseif prob_type == 3
        # Multiplication of mixed numbers
        whole1 = randint(2, 6)
        whole2 = randint(2, 6)
        denom1 = choice([2, 3, 4, 5, 6, 8])
        denom2 = choice([2, 3, 4, 5, 6, 8])
        num1 = randint(1, denom1 - 1)
        num2 = randint(1, denom2 - 1)
        
        imp1 = whole1 * denom1 + num1
        imp2 = whole2 * denom2 + num2
        
        result_num = imp1 * imp2
        result_denom = denom1 * denom2
        
        final_whole = result_num ÷ result_denom
        final_num = result_num % result_denom
        
        g = gcd(final_num, result_denom)
        final_num_simp = final_num ÷ g
        final_denom = result_denom ÷ g
        
        answer = final_num_simp == 0 ? "$final_whole" : "$(final_whole) $(final_num_simp)/$(final_denom)"
        
        problem(
            question="Multiply and simplify: \$$(whole1)\\frac{$(num1)}{$(denom1)} \\times $(whole2)\\frac{$(num2)}{$(denom2)}\$",
            answer=answer,
            difficulty=(1400, 1600),
            solution=steps(
                "Convert to improper fractions: \$\\frac{$(imp1)}{$(denom1)} \\times \\frac{$(imp2)}{$(denom2)}\$",
                "Multiply: \$\\frac{$(result_num)}{$(result_denom)}\$",
                "Simplify and convert: \$$(answer)\$"
            ),
            time=180
        )
        
    elseif prob_type == 4
        # Division of mixed numbers
        whole1 = randint(2, 7)
        whole2 = randint(1, 5)
        denom1 = choice([2, 3, 4, 5, 6])
        denom2 = choice([2, 3, 4, 5, 6])
        num1 = randint(1, denom1 - 1)
        num2 = randint(1, denom2 - 1)
        
        imp1 = whole1 * denom1 + num1
        imp2 = whole2 * denom2 + num2
        
        result_num = imp1 * denom2
        result_denom = denom1 * imp2
        
        final_whole = result_num ÷ result_denom
        final_num = result_num % result_denom
        
        g = gcd(final_num, result_denom)
        final_num_simp = final_num ÷ g
        final_denom = result_denom ÷ g
        
        answer = final_num_simp == 0 ? "$final_whole" : "$(final_whole) $(final_num_simp)/$(final_denom)"
        
        problem(
            question="Divide and simplify: \$$(whole1)\\frac{$(num1)}{$(denom1)} \\div $(whole2)\\frac{$(num2)}{$(denom2)}\$",
            answer=answer,
            difficulty=(1500, 1700),
            solution=steps(
                "Convert to improper fractions: \$\\frac{$(imp1)}{$(denom1)} \\div \\frac{$(imp2)}{$(denom2)}\$",
                "Multiply by reciprocal: \$\\frac{$(imp1)}{$(denom1)} \\times \\frac{$(denom2)}{$(imp2)}\$",
                "Multiply: \$\\frac{$(result_num)}{$(result_denom)}\$",
                "Simplify and convert: \$$(answer)\$"
            ),
            time=180
        )
        
    elseif prob_type == 5
        # Addition with unlike denominators
        whole1 = randint(1, 6)
        whole2 = randint(1, 6)
        denom1 = choice([3, 4, 6])
        denom2 = choice([2, 5, 8])
        num1 = randint(1, denom1 - 1)
        num2 = randint(1, denom2 - 1)
        
        lcd = lcm(denom1, denom2)
        new_num1 = num1 * (lcd ÷ denom1)
        new_num2 = num2 * (lcd ÷ denom2)
        
        sum_num = new_num1 + new_num2
        extra_wholes = sum_num ÷ lcd
        final_num = sum_num % lcd
        final_whole = whole1 + whole2 + extra_wholes
        
        g = gcd(final_num, lcd)
        final_num_simp = final_num ÷ g
        final_denom = lcd ÷ g
        
        answer = final_num_simp == 0 ? "$final_whole" : "$(final_whole) $(final_num_simp)/$(final_denom)"
        
        problem(
            question="Add and simplify: \$$(whole1)\\frac{$(num1)}{$(denom1)} + $(whole2)\\frac{$(num2)}{$(denom2)}\$",
            answer=answer,
            difficulty=(1400, 1600),
            solution=steps(
                "Find LCD of $(denom1) and $(denom2): LCD = $(lcd)",
                "Convert fractions: \$\\frac{$(num1)}{$(denom1)} = \\frac{$(new_num1)}{$(lcd)}\$ and \$\\frac{$(num2)}{$(denom2)} = \\frac{$(new_num2)}{$(lcd)}\$",
                "Add: \$$(whole1) + $(whole2) + \\frac{$(sum_num)}{$(lcd)}\$",
                "Simplify: \$$(answer)\$"
            ),
            time=180
        )
        
    else
        # Complex word problem with mixed numbers
        recipe_mult = choice([2, 3, 4])
        denom = choice([2, 3, 4])
        whole = randint(1, 3)
        num = randint(1, denom - 1)
        
        total_improper = recipe_mult * (whole * denom + num)
        total_whole = total_improper ÷ denom
        total_num = total_improper % denom
        
        g = gcd(total_num, denom)
        total_num_simp = total_num ÷ g
        total_denom = denom ÷ g
        
        answer = total_num_simp == 0 ? "$total_whole" : "$(total_whole) $(total_num_simp)/$(total_denom)"
        
        problem(
            question="A recipe calls for \$$(whole)\\frac{$(num)}{$(denom)}\$ cups of flour. If you want to make $(recipe_mult) batches of the recipe, how many cups of flour do you need?",
            answer=answer,
            difficulty=(1500, 1800),
            solution=steps(
                "Multiply the amount by number of batches: \$$(recipe_mult) \\times $(whole)\\frac{$(num)}{$(denom)}\$",
                "Convert to improper fraction: \$$(recipe_mult) \\times \\frac{$(whole * denom + num)}{$(denom)}\$",
                "Multiply: \$\\frac{$(total_improper)}{$(denom)}\$",
                "Convert and simplify: \$$(answer)\$ cups"
            ),
            time=200
        )
    end
end